sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/sns/SnsIO.java

```diff
@@ -62,310 +58,272 @@
-    }
-
-    /**
-     * An interface used to control if we retry the SNS Publish call when a {@link Throwable}
-     * occurs. If {@link RetryPredicate#test(Object)} returns true, {@link Write} tries to resend
-     * the requests to the Solr server if the {@link RetryConfiguration} permits it.
-     */
-    @FunctionalInterface
-    interface RetryPredicate extends Predicate<Throwable>, Serializable {}
-
-    private static class DefaultRetryPredicate implements RetryPredicate {
-      private static final ImmutableSet<Integer> ELIGIBLE_CODES =
-          ImmutableSet.of(HttpStatus.SC_SERVICE_UNAVAILABLE);
-
-      @Override
-      public boolean test(Throwable throwable) {
-        return (throwable instanceof IOException
-            || (throwable instanceof InternalErrorException)
-            || (throwable instanceof InternalErrorException
-                && ELIGIBLE_CODES.contains(((InternalErrorException) throwable).statusCode())));
-      }
-    }
-  }
-
-  /** Implementation of {@link #write}. */
+  /** Implementation of {@link #write()}. */
   @AutoValue
   public abstract static class Write<T>
-      extends PTransform<PCollection<T>, PCollection<PublishResponse>> {
-    @Nullable
-    abstract String getTopicArn();
+      extends PTransform<PCollection<T>, PCollection<WrappedSnsResponse<T>>> {
 
     @Nullable
-    abstract SerializableFunction<T, PublishRequest> getPublishRequestFn();
+    abstract SnsClientProvider getSnsClientProvider();
+
+    /** Flag to indicate if io should throw an error on exception. */
+    abstract boolean getThrowOnError();
 
+    /** Coder for element T. */
     @Nullable
-    abstract SnsClientProvider getSnsClientProvider();
+    abstract Coder<T> getElementCoder();
 
+    /** SerializableFunction to create PublishRequest. */
     @Nullable
-    abstract RetryConfiguration getRetryConfiguration();
+    abstract SerializableFunction<T, PublishRequest> getPublishRequestFn();
 
     abstract Builder<T> builder();
 
     @AutoValue.Builder
     abstract static class Builder<T> {
+      abstract Builder<T> setSnsClientProvider(SnsClientProvider clientProvider);
 
-      abstract Builder<T> setTopicArn(String topicArn);
+      abstract Builder<T> setThrowOnError(boolean throwOnError);
+
+      abstract Builder<T> setElementCoder(Coder<T> elementCoder);
 
       abstract Builder<T> setPublishRequestFn(
           SerializableFunction<T, PublishRequest> publishRequestFn);
 
-      abstract Builder<T> setSnsClientProvider(SnsClientProvider snsClientProvider);
-
-      abstract Builder<T> setRetryConfiguration(RetryConfiguration retryConfiguration);
-
       abstract Write<T> build();
     }
 
     /**
-     * Specify the SNS topic which will be used for writing, this name is mandatory.
+     * Allows to specify a flag to indicate if io should throw an error on exception.
      *
-     * @param topicArn topicArn
+     * @param throwOnError flag.
      */
-    public Write<T> withTopicArn(String topicArn) {
```

Comment:
To prevent breaking public API, please, deprecate public API methods before removing them.